### ISSUE TITLE
[FEATURE] Scheduling of VM Backups

### ISSUE THREAD

**Issue by Martin-Weiss** [harvester/harvester#2756]
**Is your feature request related to a problem? Please describe.**
It seems we can only create "manual" backups of VMs - but we need a way to schedule backups and retire / cleanup old backups.

Similar to longhorn where we can create backups per volume on a scheduled basis with the configuration option to keep them for a specific time or only keep a specific number of backups - we should have that on a Harvester / Namespace / VM level, too.

**Describe the solution you'd like**
Configuration option for global backups, namespaced backups and VM based backups to schedule.
This also should include alerting and monitoring for successful and failed backups.

There should also be an option to backup while the VM is running and include memory / snapshot / pause the VM or shutdown (gracefully), backup and then startup again..

**Describe alternatives you've considered**
Longhorn based scheduled backup - but that is per volume and not per VM so restore is a huge challenge.

P.S. the same should be there for VM snapshots once we support them.

**Comment by Martin-Weiss** [comment]
While we are waiting on v1.2.0 is there a manual workaround we could use? I.e. create a kubernetes batch job?

**Comment by bk201** [comment]
Some nice comments from @weihanglo from https://github.com/harvester/harvester/issues/572

A possible user experience may be:

- When a user want to create a recurring backup
    1. A user goes to "Cron Jobs" page on the web UI.
    2. Then the user selects "create backup" job type and a dialog shows up.
    3. The user inputs [cron format] and select a VM to backup.
    4. The backup will start in schedule.
- A user also can check job status from the "Cron Jobs" page
   - A user can read their statuses directly.
   - A user can also read information of past jobs
 
Of course, we can  leverage `Job` and `CronJob`, which are K8s resource, so we just need to add cronjob actions API. No additional CR needed. We can tweak those statuses in a user-friendly to hide some implementation details as well.

Since [CronJobs] GA in K8s 1.21, the problem is, how much flexibility do we provide to users?

[CronJobs]: https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/
[cron format]: https://en.wikipedia.org/wiki/Cron

---

That would be great if we can reuse Longhorn's [recurring job] feature by specifying longhorn volume `volume.spec.recurringJobs`. And because Longhorn already [validates recurring jobs input], we might want to patch the volume directly from Harvester web UI via RESTful API.

The questions are:

1. Do we accept highly coupling between Longhorn and Harvester web UI? Or should we introduce a new API to separate UI from implementation details? For the latter, we'll need to do input validation (including cron format) again for a true separation.
2. Usually, a VM might contain multiple volumes. In Harvester we treat custom resource `vmbackup` as a "volume group" for one-to-many VM-volume mapping. If we decide to use Longhorn recurring feature, we would need another way to maintain the relation between VM backup cronjobs and longhorn recurring jobs.

Note that we might want to specify [`AllowRecurringJobWhileVolumeDetached` setting] if using Longhorn's recurring backup feature.

Ref: https://longhorn.io/docs/1.1.2/snapshots-and-backups/scheduling-backups-and-snapshots/

[recurring job]: https://github.com/longhorn/longhorn-manager/blob/8583569b46ebd1288214cc726ace0dd54117cdb3/types/resource.go#L163-L169
[validates recurring jobs input]: https://github.com/longhorn/longhorn-manager/blob/8583569b46ebd1288214cc726ace0dd54117cdb3/manager/volume.go#L801-L832
[`AllowRecurringJobWhileVolumeDetached` setting]: https://github.com/longhorn/longhorn-manager/blob/8583569b46ebd1288214cc726ace0dd54117cdb3/controller/volume_controller.go#L2826-L2836

**Comment by abonillabeeche** [comment]
With this on 1.4, we need to solidify the strategy. I would be OK if the backup job is done via the Longhorn UI if the resulting backup shows up in Harvester. Or we copy the recurring features of Longhorn into Harvester.

**Comment by WebberHuang1118** [comment]
We should check if LH issue [#7744](https://github.com/longhorn/longhorn/issues/7744) and [#8742](https://github.com/longhorn/longhorn/issues/8742) bring any affect on this ticket

**Comment by bk201** [comment]
We need to add known issue for https://github.com/longhorn/longhorn/issues/8742

**Comment by WebberHuang1118** [comment]
> We need to add known issue for [longhorn/longhorn#8742](https://github.com/longhorn/longhorn/issues/8742)

This is expected to be solved by https://github.com/longhorn/longhorn/pull/9152

**Comment by WebberHuang1118** [comment]
Hi @TachunLin 

The testing plan has been updated https://github.com/harvester/harvester/pull/6149#issue-2399969765, please check with related GUI issue https://github.com/harvester/harvester/issues/6058

**Comment by johnwc** [comment]
Are we able to do on-demand backups from an existing scheduled backup? So they can choose an existing schedule and start it immediately.

**Comment by TachunLin** [comment]
### Test Schedule VM Backup and Snapshot type

Verified **fixed** on `v1.4-a14e133-head` (24/10/09).

### Schedule Backup Result
<details><summary>$\color{green}{\textsf{PASS}}$ Case0: Basic validation for CR "ScheduleVMBackup" $~~$</summary>

  1. Set invalid `spec.retain < 2` would be prevented

  ```
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  # schedulevmbackups.harvesterhci.io "svmbackup-vm" was not valid:
  # * spec.retain: Invalid value: 1: spec.retain in body should be greater than or equal to 2
  #
  apiVersion: harvesterhci.io/v1beta1
  kind: ScheduleVMBackup
  metadata:
    annotations:
      harvesterhci.io/svmbackupSkipCronCheck: enable
    creationTimestamp: "2024-10-10T15:00:37Z"
    finalizers:
    - wrangler.cattle.io/schedule-vm-bakcup-controller
    generation: 8154
    name: svmbackup-vm
    namespace: default
    resourceVersion: "6704800"
    uid: fa42c56f-0b02-4cb8-8e15-1817b5d7c510
  spec:
    cron: '*/2 * * * *'
    maxFailure: 2
    retain: 1
  
  ```

  2. Set invalid `spec.retain` > 250 would be prevented

  ```
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  # schedulevmbackups.harvesterhci.io "svmbackup-vm" was not valid:
  # * spec.retain: Invalid value: 251: spec.retain in body should be less than or equal to 250
  #
  apiVersion: harvesterhci.io/v1beta1
  kind: ScheduleVMBackup
  metadata:
    annotations:
      harvesterhci.io/svmbackupSkipCronCheck: enable
    creationTimestamp: "2024-10-10T15:00:37Z"
    finalizers:
    - wrangler.cattle.io/schedule-vm-bakcup-controller
    generation: 8151
    name: svmbackup-vm
    namespace: default
    resourceVersion: "6704331"
    uid: fa42c56f-0b02-4cb8-8e15-1817b5d7c510
  spec:
    cron: '*/2 * * * *'
    maxFailure: 2
    retain: 251
    suspend: false
  ```


  3. Set invalid `spec.maxFailure` < 2 would be prevented

  ```
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  # schedulevmbackups.harvesterhci.io "svmbackup-vm" was not valid:
  # * spec.maxFailure: Invalid value: 1: spec.maxFailure in body should be greater than or equal to 2
  #
  apiVersion: harvesterhci.io/v1beta1
  kind: ScheduleVMBackup
  metadata:
    annotations:
      harvesterhci.io/svmbackupSkipCronCheck: enable
    creationTimestamp: "2024-10-10T15:00:37Z"
    finalizers:
    - wrangler.cattle.io/schedule-vm-bakcup-controller
    generation: 8154
    name: svmbackup-vm
    namespace: default
    resourceVersion: "6704800"
    uid: fa42c56f-0b02-4cb8-8e15-1817b5d7c510
  spec:
    cron: '*/2 * * * *'
    maxFailure: 1
    retain: 3
    suspend: false
  
  ```

  4. Set `spec.maxFailure` >= `spec.retain` will be prevented
 
  ```
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  # schedulevmbackups.harvesterhci.io "svmbackup-vm" was not valid:
  # * spec.maxFailure: max failure should be less than retain
  #
  apiVersion: harvesterhci.io/v1beta1
  kind: ScheduleVMBackup
  metadata:
    annotations:
      harvesterhci.io/svmbackupSkipCronCheck: enable
    creationTimestamp: "2024-10-10T15:00:37Z"
    finalizers:
    - wrangler.cattle.io/schedule-vm-bakcup-controller
    generation: 8166
    name: svmbackup-vm
    namespace: default
    resourceVersion: "6710058"
    uid: fa42c56f-0b02-4cb8-8e15-1817b5d7c510
  spec:
    cron: '*/2 * * * *'
    maxFailure: 3
    retain: 3
  
  ```


 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Case1: Checking the backup rotation $~~$</summary>

  1. Apply manifest to create a scheduled VM backup
  ```
  apiVersion: harvesterhci.io/v1beta1
  kind: ScheduleVMBackup
  metadata:
    annotations:
        harvesterhci.io/svmbackupSkipCronCheck: "enable"
    namespace: default
    name: svmbackup-vm
  spec:
    cron: '*/2 * * * *'
    retain: 3
    maxFailure: 2
    vmbackup:
      source:
        apiGroup: kubevirt.io
        kind: VirtualMachine
        name: test
      type: backup
  ```
  2. Created the schedule backup
  ![image](https://github.com/user-attachments/assets/367b22ea-5dd5-4709-b5fc-adc2b9633568)

  3. Checking new backups keep being created every two minutes

  ![image](https://github.com/user-attachments/assets/950c0ee6-226f-444f-b15f-5ccf3104d937)

 4. The Harvester always maintains at least three VM backups since the 6th minute 
 
  ![image](https://github.com/user-attachments/assets/b4099dcd-f05a-41ee-b35b-6214c26613ff)
  ![image](https://github.com/user-attachments/assets/6fad5e37-333f-4f83-a1dc-798804f93703)

 
 5. The oldest VM backup will be deleted as new VM backup completes

  ![image](https://github.com/user-attachments/assets/04c9aa51-9dd2-46f8-94d7-785d1e65b8dc)
 

 6. The oldest Longhorn snapshot should also be purged under the hood

  ![image](https://github.com/user-attachments/assets/df0e81b2-c199-4411-be4b-9e3e1399c5f3)

  ![image](https://github.com/user-attachments/assets/96a45c4b-62a6-4ff3-bbdb-7cc5e8d03c50)
  ![image](https://github.com/user-attachments/assets/b1986718-1198-414e-bf8f-1ebd6743d3f5)

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Case2: Scheduled backup for a stopped VM $~~$</summary>


  1. Checking new backups on the stopped vm keep being created every two minutes
  
  ![image](https://github.com/user-attachments/assets/21ba3ede-2963-4b2d-b268-e9eb043dc3eb)

  2. Related resources should keep being updated

  ![image](https://github.com/user-attachments/assets/04e8b5de-8e86-47d9-a095-82de034578f1)
  ![image](https://github.com/user-attachments/assets/79948080-3f44-4a10-918a-c4d061d3e4a7)

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Case3: Scheduled backup keeps processing if temporal backup failure happens $~~$</summary>

1. When backup target is unreachable, check the failure backup can be discovered
  - ScheduleVMBackup CR `.status.failure` = 1

  ```
  spec:
    cron: '*/2 * * * *'
    maxFailure: 2
    retain: 3
    suspend: false
    vmbackup:
      source:
        apiGroup: kubevirt.io
        kind: VirtualMachine
        name: test
      type: backup
  status:
    failure: 1
    vmbackupInfo:
    - name: svmb-40ebc943-2afa-4a13-989e-132ed6c6d0aa-20241015.1524
      readyToUse: true
      volumeBackupInfo:
      - name: svmb-40ebc943-2afa-4a13-989e-132ed6c6d0aa-20241015.1524-volume-test-disk-0-7pcpn
  ```

 - Backups display failure result
  ![image](https://github.com/user-attachments/assets/a08bea46-ee1f-453d-9aef-7eaafd09d6d4)


2. When backup target is resumed, check new backups keep being created every two minutes

  ![image](https://github.com/user-attachments/assets/907c8526-9e30-435c-9e6f-00f55cc0f81e)
  
  ![image](https://github.com/user-attachments/assets/3987b92a-a3c0-43e2-abd8-f0835c160164)

 - `.status.failure` value would be clean up to disappear

 - Failure backup is removed
  ![image](https://github.com/user-attachments/assets/a2e186a8-ae12-4455-9b59-bc5f3816e890)
 
 - The backup schedule can resume correctly, new backups continue being created every two minutes
  ![image](https://github.com/user-attachments/assets/a7f4819e-02da-40bc-80af-daccc2ae5793)

  ```
  status:
    vmbackupInfo:
    - name: svmb-40ebc943-2afa-4a13-989e-132ed6c6d0aa-20241015.1524
      readyToUse: true
      volumeBackupInfo:
      - name: svmb-40ebc943-2afa-4a13-989e-132ed6c6d0aa-20241015.1524-volume-test-disk-0-7pcpn
    - name: svmb-40ebc943-2afa-4a13-989e-132ed6c6d0aa-20241015.1528
      readyToUse: true
      volumeBackupInfo:
      - name: svmb-40ebc943-2afa-4a13-989e-132ed6c6d0aa-20241015.1528-volume-test-disk-0-7pcpn
    - name: svmb-40ebc943-2afa-4a13-989e-132ed6c6d0aa-20241015.1530
      readyToUse: true
      volumeBackupInfo:
      - name: svmb-40ebc943-2afa-4a13-989e-132ed6c6d0aa-20241015.1530-volume-test-disk-0-7pcpn
  ```

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Case4: Failure backups reach threshold $~~$</summary>

  1. When the backup target is unreachable and backup hits max failure threshold 

  ```
  status:
    conditions:
    - lastUpdateTime: "2024-10-16T02:44:39Z"
      status: "False"
      type: BackupSuspend
    failure: 2
    vmbackupInfo:
  
  ```

  2. The schedules backup should be suspended

  ```
  spec:
    cron: '*/2 * * * *'
    maxFailure: 2
    retain: 3
    suspend: true
    vmbackup:
      source:
        apiGroup: kubevirt.io
        kind: VirtualMachine
        name: test
      type: backup
  ```

  ![image](https://github.com/user-attachments/assets/d4d7f33e-8c73-4630-8804-8c126698321e)


  3. No more new VM backup is going to be created in this phase

  ![image](https://github.com/user-attachments/assets/39276364-de8d-408f-8469-d97929a092cd)


 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Case5: Resuming suspended backup schedule $~~$</summary>

  1. When the backup target `connection recovered` and we click the resume on the schedule backup

  ![image](https://github.com/user-attachments/assets/b6fb9c96-0c2b-4449-80fe-58661535f18e)


  2. The schedules backup should be resumed

  ![image](https://github.com/user-attachments/assets/077dd329-1924-4b4c-8c37-c7a80aeedb44)
  
  ```
  spec:
    cron: '*/2 * * * *'
    maxFailure: 2
    retain: 3
    suspend: false
    vmbackup:
      source:
        apiGroup: kubevirt.io
        kind: VirtualMachine
        name: test
      type: backup
  ```



  3. Can continue with the backup and  clean the failed backup

  ![image](https://github.com/user-attachments/assets/bd762044-f1ff-40e9-9fd0-6961e66ea8ee)

  ![image](https://github.com/user-attachments/assets/c363d0f7-ff44-423c-a8fe-b2408a799747)


 </details>

### Schedule Snapshot Result

<details><summary>$\color{green}{\textsf{PASS}}$ Case1: Checking the snapshot rotation $~~$</summary>

  1. Apply manifest to create a scheduled VM backup

  ```
  apiVersion: harvesterhci.io/v1beta1
  kind: ScheduleVMBackup
  metadata:
    annotations:
        harvesterhci.io/svmbackupSkipCronCheck: "enable"
    namespace: default
    name: svmbackup-snap
  spec:
    cron: '*/2 * * * *'
    retain: 3
    maxFailure: 2
    vmbackup:
      source:
        apiGroup: kubevirt.io
        kind: VirtualMachine
        name: vm
      type: snapshot
  ```

  2. Created the schedule snapshot for vm

  ![image](https://github.com/user-attachments/assets/ff5852d4-8ac5-46a9-84bc-197412c15c0d)


  3. Checking new snapshots keep being created every two minutes

  ![image](https://github.com/user-attachments/assets/5fed30aa-0113-4338-935e-f10a99d5e27f)


 4. The Harvester always maintains at least three VM snapshot since the 6th minute 
 
  ![image](https://github.com/user-attachments/assets/cbf89400-a5fc-41cf-87cb-303e64541c11)


 5. The oldest VM snapshot will be deleted as new VM snapshot completes

![image](https://github.com/user-attachments/assets/3d9f72a1-6430-4f67-b005-a56b9fccaaee)

![image](https://github.com/user-attachments/assets/d57a1ce7-c209-43d8-97ba-74be7f73f1b2)


 6. The oldest Longhorn snapshot should also be purged under the hood

![image](https://github.com/user-attachments/assets/da8dfce5-b55a-4c31-ac0b-9d61f6cbb105)

![image](https://github.com/user-attachments/assets/739dc5d8-121c-404e-b7cd-6e5518a0f2fe)


 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Case2: ScheduleVMBackup for snapshot wouldn't be rejected even if the backup target is not reachable $~~$</summary>


  1. Can created the schedule snapshot for vm when the backup target is unreachable
  
  ![image](https://github.com/user-attachments/assets/ff5852d4-8ac5-46a9-84bc-197412c15c0d)


 </details>





 </details>

### Test Information
* Test Environment: Single nodes harvester on equinix bare machines
* Harvester version: `v1.4-a14e133-head` (24/10/09)

### Verify Steps
<details><summary> Prerequisite </summary>

1. Prepare a VM or machine with available connection with Harvester
2. Use docker to setup Minio server 
  ```
  $ mkdir /data/volume-s3
  $ chmod 777 /data/volume-s3
   
  $ docker run -d --name minio \
  --restart=unless-stopped \
  -e MINIO_ROOT_USER=minioadmin \
  -e MINIO_ROOT_PASSWORD=minioadmin \
  -v /data/volume-s3:/data \
  -p 9000:9000 \
  -p 9001:9001 \
  bitnami/minio:latest
   
  $ docker ps | grep minio
  ff3909992298   bitnami/minio:latest 
  ```
3. Access the minio endpoint http://192.168.0.254:9000, input the MINIO_ROOT_USER and MINIO_ROOT_PASSWORD to login
4. Set the region `pcloud` in configuration
5. Create a bucket `mybucket`
6. Set the S3 backup target in Harvester
  ```
  {
    "type": "s3",
    "endpoint": "http://192.168.0.254:9000",
    "accessKeyId": "",
    "secretAccessKey": "",
    "bucketName": "mybucket",
    "bucketRegion": "pcloud",
    "cert": "",
    "virtualHostedStyle": false
  }
  ```

</details>
<details><summary> VM schedule backup steps </summary>

refer to: https://github.com/harvester/harvester/pull/6149#issue-2399969765

</details>

**Comment by WebberHuang1118** [comment]
> ### Test Schedule Backup Type
> Verifying on `v1.4-a14e133-head` (24/10/09).
> 
> ### Result
> Failed
> 4. ❌ Can still set `spec.maxFailure.default` < `spec.retain.default` and it's the default Harvester setting
> 
> ```
>           spec:
>             properties:
>               cron:
>                 type: string
>               maxFailure:
>                 default: 4
>                 minimum: 2
>                 type: integer
>               retain:
>                 default: 8
>                 maximum: 250
>                 minimum: 2
>                 type: integer
> ```

Hi @TachunLin 

The CR format is `.spec.maxFailure` and `.spec.retain`, there is no field for `.spec.maxFailure.default/maximum/minimum` and `.spec.retain.default/maximum/minimum` 

thanks.

**Comment by TachunLin** [comment]
Thanks @WebberHuang1118 for the reminder. 

It should check the `ScheduleVMBackup` CR rather then the CRD. 
There should be able to set value to `spec.masFailure` and `spec.retain`. I have updated the test result accordingly.

**Comment by TachunLin** [comment]
### Test VM Backup Schedule Operations

Verified **fixed** on `v1.4-a14e133-head` (24/10/09).

### Duplicated Schedule for the same VM result
<details><summary>$\color{green}{\textsf{PASS}}$ Case1: Two schedule backups on the same vm  $~~$</summary>

  1. When creating two CR `ScheduleVMBackup` backup targets on the same VM, it will be `rejected` 

  ```
  apiVersion: harvesterhci.io/v1beta1
  kind: ScheduleVMBackup
  metadata:
    annotations:
        harvesterhci.io/svmbackupSkipCronCheck: "enable"
    namespace: default
    name: svmbackup2-vm
  spec:
    cron: '*/2 * * * *'
    retain: 3
    maxFailure: 2
    vmbackup:
      source:
        apiGroup: kubevirt.io
        kind: VirtualMachine
        name: test
      type: backup
  ```

  ```
  harvester-node-1:~ # kubectl create -f schedule-vm-backup2.yaml
  The request is invalid: spec.vmbackup: VM default/test already has backup schedule
  ```


 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Case2: Another schedule snapshot on the same vm already have backup schedule $~~$</summary>

  1.When creating another CR `ScheduleVMBackup` snapshot targets to on the same VM already have backup schedule, it will be `rejected` 


  ```
  apiVersion: harvesterhci.io/v1beta1
  kind: ScheduleVMBackup
  metadata:
    annotations:
        harvesterhci.io/svmbackupSkipCronCheck: "enable"
    namespace: default
    name: svmbackup3-backup
  spec:
    cron: '*/2 * * * *'
    retain: 3
    maxFailure: 2
    vmbackup:
      source:
        apiGroup: kubevirt.io
        kind: VirtualMachine
        name: vm
      type: backup
  ```

  ```
  harvester-node-1:~ # kubectl create -f schedule-vm-backup3.yaml
  The request is invalid: spec.vmbackup: VM default/vm already has backup schedule
  ```

 </details>

### Proactive Suspend/Resume the Schedule 

<details><summary>$\color{green}{\textsf{PASS}}$ Case1: Edit to suspend the ScheduleVMBackup $~~$</summary>

  1. After setting the `spec.suspend`  = true, no more new VM backups will be created 

  ```
  spec:
    cron: '*/2 * * * *'
    maxFailure: 2
    retain: 3
    suspend: true
    vmbackup:
      source:
        apiGroup: kubevirt.io
        kind: VirtualMachine
        name: test
      type: backup
  ```

  ![image](https://github.com/user-attachments/assets/1059bde6-b630-49ae-8eb5-95e41a8e9140)


 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Case2: Edit to resume the ScheduleVMBackup $~~$</summary>

  1. After setting the `spec.suspend` = false, new VM backups will be created in every two minutes

  ```
  spec:
    cron: '*/2 * * * *'
    maxFailure: 2
    retain: 3
    suspend: false
    vmbackup:
      source:
        apiGroup: kubevirt.io
        kind: VirtualMachine
        name: test
      type: backup
  ```

  ![image](https://github.com/user-attachments/assets/391131bf-54e8-443e-8f10-b98fb08298db)
  
  
  ![image](https://github.com/user-attachments/assets/507a2fab-795b-483c-8b97-a0d1f9e48c73)


 </details>

###  Runtime update schedule cron expression

<details><summary>$\color{green}{\textsf{PASS}}$ Case1: Update the retain and max failure value of schedule backup resource $~~$</summary>

  1. Increase the `spec.retain` to 4, check can retain at most 4 vm backups

  ![image](https://github.com/user-attachments/assets/59dc8552-364f-481b-ac92-b91c18926c5e)

  ![image](https://github.com/user-attachments/assets/1d766d2d-aefd-4d17-98c3-96be34e7f19a)


  2.  Increase the `spec.maxFailure` to 3, check can only accept at most 3 failed backups

  ![image](https://github.com/user-attachments/assets/dffcb0b2-b293-40c9-93fa-f2bd97ad06c5)

  ![image](https://github.com/user-attachments/assets/167d82da-c594-4ca2-b9a2-8151058864fb)

  ![image](https://github.com/user-attachments/assets/ee84d222-c1c8-4ef8-abc7-94f76fc2df3e)

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Case2: Update the cron job value of schedule backup resource $~~$</summary>

  1. After update the cron job value to `spec.cron = */3 * * * *`, the vm will be created in every three minutes

  ![image](https://github.com/user-attachments/assets/a584ca7a-4f91-48af-a804-ae28173b1dbf)

  ![image](https://github.com/user-attachments/assets/5ce7a576-414c-40c2-aaa6-ff29c54bdbc5)


 </details>


###  Schedule Granularity Check

<details><summary>$\color{green}{\textsf{PASS}}$ Case1: If the schedule's granularity is less than one hour, it can't be applied $~~$</summary>

  1. When you did not specify the `svmbackupSkipCronCheck` in the scheduled backup and set the cron time less than **one hour**

  ```
  apiVersion: harvesterhci.io/v1beta1
  kind: ScheduleVMBackup
  metadata:
    namespace: default
    name: svmbackup-vm1
  spec:
    cron: '*/2 * * * *'
    retain: 3
    maxFailure: 2
    vmbackup:
      source:
        apiGroup: kubevirt.io
        kind: VirtualMachine
        name: vm1
      type: backup
  ```

  2.  The creation would be ejected by webhook with message: `The request is invalid: spec.cron: schedule granularity `2m0s` less than `1h0m0s``

  ```
  harvester-node-1:~ # kubectl create -f svmbackup-vm1.yaml
  The request is invalid: spec.cron: schedule granularity 2m0s less than 1h0m0s
  ```


 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Case2: If the schedule's granularity is equal or more than one hour, it could be applied $~~$</summary>

  1. When you disable the `svmbackupSkipCronCheck` in the scheduled backup and set the cron time more than **one hour**

  ```
  apiVersion: harvesterhci.io/v1beta1
  kind: ScheduleVMBackup
  metadata:
    namespace: default
    name: svmbackup-vm1
  spec:
    cron: '0 * * * *'
    retain: 3
    maxFailure: 2
    vmbackup:
      source:
        apiGroup: kubevirt.io
        kind: VirtualMachine
        name: vm1
      type: backup
  ```


  2.  The creation would be success

  ```
  harvester-node-1:~ # kubectl create -f svmbackup-vm1.yaml
  schedulevmbackup.harvesterhci.io/svmbackup-vm1 created
  ```

  ![image](https://github.com/user-attachments/assets/41edb1de-af90-4b05-835e-6af5ea0b7194)


 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Case3:  If two schedules have the same granularity and each iteration's timing offset is less than 10mins, the creation would be ejected $~~$</summary>

  1. When you disable the `svmbackupSkipCronCheck` in the scheduled backup and already have a schedule backup at `0 * * * *`

  2. After you create another schedule backup resource at `5 * * * *`

  ```
  apiVersion: harvesterhci.io/v1beta1
  kind: ScheduleVMBackup
  metadata:
    namespace: default
    name: svmbackup-vm2
  spec:
    cron: '5 * * * *'
    retain: 3
    maxFailure: 2
    vmbackup:
      source:
        apiGroup: kubevirt.io
        kind: VirtualMachine
        name: vm2
      type: backup
  ```

  
  3. The creation would be ejected by webhook with The request is invalid: spec.cron: same granularity as schedule svmbackup-vm1 and offset less than 10m0s 

  ```
  harvester-node-1:~ # kubectl create -f svmbackup-vm2.yaml
  The request is invalid: spec.cron: same granularity as schedule svmbackup-vm1 and offset less than 10m0s
  ```


 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Case4:  If two schedules have the same granularity and each iteration's timing offset is <strong>not less than</strong> 10mins, the creation could be applied $~~$</summary>

  1. When you disable the `svmbackupSkipCronCheck` in the scheduled backup and already have a schedule backup at `0 * * * *`

  2. After you create another schedule backup resource at `10 * * * *`

  ```
  apiVersion: harvesterhci.io/v1beta1
  kind: ScheduleVMBackup
  metadata:
    namespace: default
    name: svmbackup-vm2
  spec:
    cron: '10 * * * *'
    retain: 3
    maxFailure: 2
    vmbackup:
      source:
        apiGroup: kubevirt.io
        kind: VirtualMachine
        name: vm2
      type: backup
  ```
  
  3. The creation should be a success
  
  ```
  harvester-node-1:~ # kubectl create -f svmbackup-vm2.yaml
  schedulevmbackup.harvesterhci.io/svmbackup-vm2 created
  ```

 </details>

### Test Information
* Test Environment: Single nodes harvester on equinix bare machines
* Harvester version: `v1.4-a14e133-head` (24/10/09)

### Verify Steps
<details><summary> Prerequisite </summary>

1. Prepare a VM or machine with available connection with Harvester
2. Use docker to setup Minio server 
  ```
  $ mkdir /data/volume-s3
  $ chmod 777 /data/volume-s3
   
  $ docker run -d --name minio \
  --restart=unless-stopped \
  -e MINIO_ROOT_USER=minioadmin \
  -e MINIO_ROOT_PASSWORD=minioadmin \
  -v /data/volume-s3:/data \
  -p 9000:9000 \
  -p 9001:9001 \
  bitnami/minio:latest
   
  $ docker ps | grep minio
  ff3909992298   bitnami/minio:latest 
  ```
4. Access the minio endpoint http://192.168.0.254:9000, input the MINIO_ROOT_USER and MINIO_ROOT_PASSWORD to login
5. Set the region `pcloud` in configuration
6. Create a bucket `mybucket`
7. Set the S3 backup target in Harvester
  ```
  {
    "type": "s3",
    "endpoint": "http://192.168.0.254:9000",
    "accessKeyId": "",
    "secretAccessKey": "",
    "bucketName": "mybucket",
    "bucketRegion": "pcloud",
    "cert": "",
    "virtualHostedStyle": false
  }
  ```

</details>
<details><summary> VM schedule backup operations </summary>

refer to: https://github.com/harvester/harvester/pull/6149#issue-2399969765

</details>

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.4-a14e133-head` (24/10/09). 

Since we have verified can pass all the test scenarios in the sections:
1. [Test functionality of Schedule VM Backup and Snapshot type](https://github.com/harvester/harvester/issues/2756#issuecomment-2405591736)
2. [Test VM Backup Schedule Operations](https://github.com/harvester/harvester/issues/2756#issuecomment-2415946187)

Thus close this issue as completed verification. Thank you.
